sycl/include/CL/sycl/handler.hpp

```diff
@@ -729,6 +810,305 @@ class __SYCL_EXPORT handler {
+      // can be split later into two separate kernels, if there are users who
+      // really need more efficient code for them.
+      using AuxName = typename detail::get_reduction_main_2nd_kernel_name_t<
+          KernelName, KernelType>::name;
+      parallel_for<AuxName>(Range, [=](nd_item<Dims> NDIt) {
+        // Call user's functions. Reducer.MValue gets initialized there.
+        typename Reduction::reducer_type Reducer(ReduIdentity);
+        KernelFunc(NDIt, Reducer);
+
+        size_t WGSize = NDIt.get_local_range().size();
+        size_t LID = NDIt.get_local_linear_id();
+        size_t GID = NDIt.get_global_linear_id();
+        // Copy the element to local memory to prepare it for tree-reduction.
+        LocalReds[LID] = (GID < NWorkItems) ? Reducer.MValue : ReduIdentity;
+        LocalReds[WGSize] = ReduIdentity;
+        NDIt.barrier();
+
+        // Tree-reduction: reduce the local array LocalReds[:] to LocalReds[0]
+        // LocalReds[WGSize] accumulates last/odd elements when the step
+        // of tree-reduction loop is not even.
+        typename Reduction::binary_operation BOp;
+        size_t PrevStep = WGSize;
+        for (size_t CurStep = PrevStep >> 1; CurStep > 0; CurStep >>= 1) {
+          if (LID < CurStep)
+            LocalReds[LID] = BOp(LocalReds[LID], LocalReds[LID + CurStep]);
+          else if (LID == CurStep && (PrevStep & 0x1))
+            LocalReds[WGSize] = BOp(LocalReds[WGSize], LocalReds[PrevStep - 1]);
+          NDIt.barrier();
+          PrevStep = CurStep;
+        }
+
+        // Compute the the partial sum/reduction for the work-group.
+        if (LID == 0) {
+          auto GrID = NDIt.get_group_linear_id();
+          auto V = BOp(LocalReds[0], LocalReds[WGSize]);
+          Out.get_pointer().get()[GrID] =
+              IsUpdateOfUserAcc ? BOp(*(Out.get_pointer()), V) : V;
+        }
+      });
+    }
+  }
+
+  /// Implements a command group function that enqueues a kernel that does one
+  /// iteration of reduction of elements in each of work-groups.
+  /// This version uses tree-reduction algorithm to reduce elements in each
+  /// of work-groups. At the end of each work-groups the partial sum is written
+  /// to a global buffer.
+  ///
+  /// Briefly: aux kernel, tree-reduction, CUSTOM types/ops.
+  template <typename KernelName, typename KernelType, int Dims, class Reduction>
+  void reduAuxCGFunc(const nd_range<Dims> &Range, size_t NWorkItems,
+                     size_t KernelRun, Reduction &Redu) {
+    size_t WGSize = Range.get_local_range().size();
+    size_t NWorkGroups = Range.get_group_range().size();
+
+    // The last work-group may be not fully loaded with work, or the work group
+    // size may be not power of those. Those two cases considered inefficient
+    // as they require additional code and checks in the kernel.
+    bool IsUnderLoaded = NWorkGroups * WGSize != NWorkItems;
+    size_t InefficientCase = (IsUnderLoaded || (WGSize & (WGSize - 1))) ? 1 : 0;
```

Comment:
It took me a while to work out why this is declared as a `size_t`.

Do you think it would be clearer if `InefficientCase` was a `bool`?  You could shift the logic of whether you need to add 1 or 0 to line 945.

Reply:
Yes, I did that. Thank you.